src/products/components/ProductUpdatePage/formChannels.ts

```diff
@@ -0,0 +1,85 @@
+import { ChannelOpts } from "@saleor/components/ChannelsAvailabilityCard/types";
+import {
+  ProductChannelListingUpdateInput,
+  ProductFragment,
+} from "@saleor/graphql";
+import useStateFromProps from "@saleor/hooks/useStateFromProps";
+import uniq from "lodash/uniq";
+import uniqBy from "lodash/uniqBy";
+import { useCallback, useRef } from "react";
+
+import { ProductChannelsListingDialogSubmit } from "./ProductChannelsListingsDialog";
+
+const emptyListing = {
+  availableForPurchaseDate: null,
+  isAvailableForPurchase: false,
+  isPublished: false,
+  publicationDate: null,
+  visibleInListings: false,
+};
+
+export function useProductChannelListingsForm(
+  product: Pick<ProductFragment, "channelListings">,
+  triggerChange: () => void,
+) {
+  const [channels, setChannels] = useStateFromProps<
```

Comment:
another useStateFromProps with complex objects to deep compare

Reply:
I won't make refactors here as this feature is huge by itself